### DISCUSSION TITLE
i2c scanning : howto ?

### DISCUSSION THREAD

**Discussion by manes6969** [tabemann/zeptoforth#47] [General]
Hi ,
I want to program an i2c scanner , that detects which i2c devices are connected on the i2c bus.
I looked in the manual (epub) , but can not find something that I can use (at least I think so).
I was thinking of sending consecutive adresses and check if there was an ack from the slave.
Is there a way of doing this in zeptoforth ?
Kind regards,
Ronny

**Comment by tabemann** [comment]
There are obvious bugs in the implementation of I2C in zeptoforth that prevent the implementation of an I2C scanner, but I have been trying to fix them to no avail. What I have been observing is after three transmits that fill up the I2C TX FIFO, triggering a NACK ─ or which do not fill up the I2C TX FIFO, and are registered as transmits (which should be a bug) ─ all transmission simply hangs, resulting in subsequent transmission attempts simply timing out. With these subsequent transmission attempts the `TX_EMPTY` interrupt is never triggered even when I explicitly clear out the TX FIFO, disable, and re-enable the I2C peripheral. One way or another, just as an example, here is my test code (note that it does not work):

```
pin import
i2c import

false value prev-enabled?

: test-i2c-addr { addr pin0 pin1 i2c-periph -- success? }
  i2c-periph clear-i2c
  prev-enabled? if
    i2c-periph disable-i2c
  then
  i2c-periph pin0 i2c-pin
  i2c-periph pin1 i2c-pin
  i2c-periph master-i2c
  addr $80 u< if i2c-periph 7-bit-i2c-addr else i2c-periph 10-bit-i2c-addr then
  addr i2c-periph ['] i2c-target-addr! try if 2drop exit then
  i2c-periph enable-i2c
  true to prev-enabled?
  task::timeout @ { old-timeout }
  1000 task::timeout !
  0. { D^ buf } buf 8 i2c-periph ['] >i2c-stop try
  dup 0= if
    drop
    cr ." ADDR: " addr . ." BYTES: " dup .
  else
    dup ['] task::x-timed-out = if
      cr ." ADDR: " addr . ." TIMED OUT"
      2drop 2drop
    else
      dup ['] x-i2c-target-noack = if
        cr ." ADDR: " addr . ." NOT ACKNOWLEDGED "
        2drop 2drop
      else
        old-timeout task::timeout !
        ?raise
      then
    then
  then
  old-timeout task::timeout !
;

: scan-i2c { pin0 pin1 i2c-periph -- }
  $400 $000 ?do i pin0 pin1 i2c-periph test-i2c-addr loop
;
```

**Comment by tabemann** [comment]
I have implemented an I2C scanner that works with the latest code in the `master` and `devel` branches (note that this code is not yet in a build). To use this code you will have to build the `full` or `full_usb` builds for the `rp2040` platform yourself until I get a new release out:

```
\ Copyright (c) 2023 Travis Bemann
\ 
\ Permission is hereby granted, free of charge, to any person obtaining a copy
\ of this software and associated documentation files (the "Software"), to deal
\ in the Software without restriction, including without limitation the rights
\ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
\ copies of the Software, and to permit persons to whom the Software is
\ furnished to do so, subject to the following conditions:
\ 
\ The above copyright notice and this permission notice shall be included in
\ all copies or substantial portions of the Software.
\ 
\ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
\ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
\ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
\ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
\ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
\ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
\ SOFTWARE.

begin-module i2c-scanner

  pin import
  i2c import
  
  : test-i2c-addr { addr pin0 pin1 i2c-periph -- }
    i2c-periph i2c::clear-i2c
    i2c-periph pin0 i2c::i2c-pin
    i2c-periph pin1 i2c::i2c-pin
    i2c-periph i2c::master-i2c
    addr $80 u< if i2c-periph 7-bit-i2c-addr else i2c-periph 10-bit-i2c-addr then
    addr i2c-periph ['] i2c::i2c-target-addr! try 0= if
      i2c-periph i2c::enable-i2c
      0. { D^ buf } buf 1 i2c-periph ['] i2c::>i2c-stop try
      i2c-periph i2c::disable-i2c
      addr h.4 ." : "
      ?dup if
        execute 2drop drop
      else
        drop ." FOUND" cr
      then
    else
      2drop
    then
  ;
  
  : scan-i2c { pin0 pin1 i2c-periph -- }
    cr $400 $000 ?do i pin0 pin1 i2c-periph test-i2c-addr loop 
  ;
  
end-module
```

To use it issue, e.g. `14 15 1 i2c-scanner::scan-i2c` if you are using I2C1 with GPIO pins 14 and 15.

**Comment by manes6969** [comment]
Hi tabemann ,
Thanks for the effort !
Since I'm not so familiar with building code , I think I will wait for the next release.
I see there is some info on github on how to do this , so I can give it a try , but don't know how this will end ...
Kind regards,
Ronny
